signac/contrib/job.py

```diff
@@ -403,7 +403,7 @@ def statepoint(self):
             you can access a dict copy of the state point by calling it, e.g.
             ``sp_dict = job.statepoint()`` instead of ``sp = job.statepoint``.
             For more information, see
-            :class:`~signac.synced_collections.backends.collection_json.JSONAttrDict`.
+            :class:`~signac._synced_collections.backends.collection_json.JSONAttrDict`.
```

Comment:
This class reference won't work because we're removing the synced collection docs. I wonder if we should add docs for the publicly imported name `signac.JSONDict` and point to that instead.

Reply:
Won't the docs for JSONDict (which is buffered) and JSONAttrDict be different?

Reply:
😕 This is tricky. We want to just point to public docs for a standalone synced collections package, but if we make synced collections hidden, we can't point to these docs.

Maybe this entire approach is flawed and we should leave `signac.synced_collections` as a public subpackage until we make `synced_collections` standalone.

cc: @glotzerlab/signac-committers

Reply:
See [my comment on the corresponding issue](https://github.com/glotzerlab/signac/issues/790#issuecomment-1229598098). Even if `synced_collections` stays public, I think pointing to a specific class here is a flawed approach. From a user's perspective, do they need any information other than the fact that this attribute is a dict-like object? If we _really_ want to explain the syncing, I would say that we could [point to the parent `SyncedDict` class](https://github.com/glotzerlab/signac/blob/master/signac/synced_collections/data_types/synced_dict.py#L24), but IMO that's exposing implementation details because synchronization is transparent.

There is only one minor exception that comes to my mind immediately, which is that the document currently supports buffering on a per-document basis:
```
with job.doc.buffered:
    # All I/O will be buffered in this block
```
but TBH I'm not sure that's something that we really need to promise at the signac level. The `signac.buffered` API should be the only entrypoint for buffering for signac users IMO:
```
with signac.buffered():
    # Buffer everything
```
since that abstracts over the details of specific backends and all instances. That abstraction guarantees that in a world where signac starts supporting different backends, users still have a single switch in the API to toggle when reads and writes are maximally buffered, and it's up to the implementation of signac to actually perform that buffering to the best of its ability. So I think that ultimately the only API promise (and therefore all that we should document) is that `[job|project].[sp|doc]` are dict-like objects, and perhaps that they support attribute-based access. We shouldn't link to a type because we don't actually promise the type of the object (again, see https://github.com/glotzerlab/signac/issues/790#issuecomment-1229598098 for why).

Reply:
I agree with this approach. Thanks @vyasr. @kidrahahjo does this make sense on what to change? If not, I can take a look at changing it myself and push to this PR.

Reply:
As far as I understood this, we do not need to expose any of the implementation details so we can either not specify the type or use MutableMappings (which is another way of saying its a dict like object) instead?